pkg/crypto/rings/cache_test.go

```diff
@@ -0,0 +1,173 @@
+			appAccount:        newAccount("secp256k1"),
+			delegateeAccounts: []account{newAccount("ed25519"), newAccount("ed25519")},
+			expectedSize:      0,
+			expectedErr:       ErrRingsWrongCurve,
+		},
+		{
+			name:              "failure: application not found",
+			appAccount:        newAccount("secp256k1"),
+			delegateeAccounts: []account{newAccount("secp256k1")},
+			expectedSize:      0,
+			expectedErr:       apptypes.ErrAppNotFound,
+		},
+	}
+	ctx := context.TODO()
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// If we expect the application to exist then add it to the test
+			// application map with the number of delegated gateways it is
+			// supposed to have so it can be retrieved from the mock
+			if !errors.As(test.expectedErr, &apptypes.ErrAppNotFound) {
+				accMap := make(map[string]cryptotypes.PubKey)
+				for _, acc := range test.delegateeAccounts {
+					accMap[acc.address] = acc.pubKey
+				}
+				// add the application's account and the accounts of all its
+				// delegated gateways to the testing state
+				testqueryclients.AddAddressToApplicationMap(t, test.appAccount.address, test.appAccount.pubKey, accMap)
+			}
+			// Attempt to retrieve the ring for the address
+			ring, err := rc.GetRingForAddress(ctx, test.appAccount.address)
+			if test.expectedErr != nil {
+				require.ErrorAs(t, err, &test.expectedErr)
+				return
+			}
+			require.NoError(t, err)
+			// Ensure the ring is the correct size.
+			require.Equal(t, test.expectedSize, ring.Size())
+		})
+	}
+}
+
+func TestRingCache_BuildRing_Cached(t *testing.T) {
+	rc := createRingCache(t)
+	tests := []struct {
+		name         string
+		appAccount   account
+		expectedSize int
+		expectedErr  error
+	}{
+		{
+			name:         "success: cached application no delegated gateways",
+			appAccount:   newAccount("secp256k1"),
+			expectedSize: 2,
+			expectedErr:  nil,
+		},
+		{
+			name:         "success: cached application with delegated gateways",
+			appAccount:   newAccount("secp256k1"),
+			expectedSize: 3,
+			expectedErr:  nil,
+		},
+	}
+	ctx := context.TODO()
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			accMap := make(map[string]cryptotypes.PubKey)
+			// if the test expects a ring > 2 we have delegated gateways
+			if test.expectedSize > 2 {
+				// create accounts for all the expected delegated gateways
+				// and add them to the map
+				for i := 0; i < test.expectedSize-1; i++ {
+					gatewayAcc := newAccount("secp256k1")
+					accMap[gatewayAcc.address] = gatewayAcc.pubKey
+				}
+			}
+			// add the application's account and the accounts of all its
+			// delegated gateways to the testing state
+			testqueryclients.AddAddressToApplicationMap(t, test.appAccount.address, test.appAccount.pubKey, accMap)
+			// Attempt to retrieve the ring for the address and cache it
+			ring1, err := rc.GetRingForAddress(ctx, test.appAccount.address)
```

Comment:
Can you add another assertion statement on the ring size BEFORE we cache it to make sure that it increases w/ the operation below.

Reply:
You have a `require.Equal(t, test.expectedSize, ring1.Size())` below, so there must be some sort of `require.GreaterThan(t, test.expectedSize, ring1.Size())`

Reply:
We cant test this until cache invalidation is in this just checks the cache caches the ring correctly

Reply:
`the cache caches the ring correctly` - Why not?